### ISSUE TITLE
plugins/tor: python3, pylint

### ISSUE THREAD

**Issue by ilf** [munin-monitoring/contrib#1061]
[plugins/tor/tor_](https://raw.githubusercontent.com/munin-monitoring/contrib/master/plugins/tor/tor_) currently has "python3" in the shebang.

For me, it only works nicely with python2.

Running [Pylint](https://pylint.org/) in "Python 3 porting mode" on [plugins/tor/tor_](https://raw.githubusercontent.com/munin-monitoring/contrib/master/plugins/tor/tor_) gives:

```
% pylint --py3k tor_
************* Module tor_
tor_:58:0: W1618: import missing `from __future__ import absolute_import` (no-absolute-import)
tor_:137:24: W1620: Calling a dict.iter*() method (dict-iter-method)
tor_:140:33: W1620: Calling a dict.iter*() method (dict-iter-method)
tor_:141:28: W1620: Calling a dict.iter*() method (dict-iter-method)
tor_:256:40: W1620: Calling a dict.iter*() method (dict-iter-method)

------------------------------------------------------------------
Your code has been rated at 9.83/10 (previous run: 9.83/10, +0.00)
```

**Comment by daftaupe** [comment]
Hello @ilf,

I found there are several places where the old iteritems is used instead of items. I guess the commit should have changed that as well.
I'm going to fix this soon :)

**Comment by ilf** [comment]
Thanks.

1. Running Pylint against https://raw.githubusercontent.com/daftaupe/contrib/tor/plugins/tor/tor_ gives:

```
% pylint --py3k tor_
************* Module tor_
tor_:58:0: W1618: import missing `from __future__ import absolute_import` (no-absolute-import)

------------------------------------------------------------------
Your code has been rated at 9.97/10 (previous run: 9.83/10, +0.13)
```

2. It still has python3 in shebang, breaking python2 environments. Probably this commit should be reverted for now: https://github.com/munin-monitoring/contrib/commit/a3319a6dc92fc25dbe6f98eab684b702befbbfff

3. I get this error:

```
Error output from tor_traffic:
       File "/etc/munin/plugins/tor_traffic", line 105
         print("Authentication failed (incorrect password)", file=sys.stderr)
                                                                 ^
     SyntaxError: invalid syntax
Service 'tor_traffic' exited with status 1/0.

```

**Comment by sumpfralle** [comment]
> I found there are several places where the old iteritems is used instead of items. I guess the commit should have changed that as well.

Are you sure, that the other mentions of `iteritems` occurred in python3 plugins? Here I just found the `tor_` plugin with this issue.


> There are still issues with the Stem library (getting fingerprint errors) but the dict ones should be fixed in https://github.com/daftaupe/contrib/tree/tor.

I am looking forward to your pull request :)


> W1618: import missing `from __future__ import absolute_import` (no-absolute-import)

The `--py3k` flag for pylint enables *porting checks*. One of these is related to absolute imports (see [some more details](https://github.com/PyCQA/pylint/issues/2115#issuecomment-389810192)). For proper testing of python3 code (already ported) you should omit the `--py3k` flag.

> It still has python3 in shebang, breaking python2 environments. Probably this commit should be reverted for now: a3319a6

I am sorry: due to python2 being end-of-life since the beginning of this year, I would strongly prefer to fix it for python3 instead.

> I get this error:

Python2 does not support the `file` keyword argument for `print`.

**Comment by daftaupe** [comment]
> > I found there are several places where the old iteritems is used instead of items. I guess the commit should have changed that as well.
> 
> Are you sure, that the other mentions of `iteritems` occurred in python3 plugins? Here I just found the `tor_` plugin with this issue.
> 
I meant in the tor_ plugin code, there are several places where iteritems was used. 

> > There are still issues with the Stem library (getting fingerprint errors) but the dict ones should be fixed in https://github.com/daftaupe/contrib/tree/tor.
> 
> I am looking forward to your pull request :)
> 

It's on its way :)

> > W1618: import missing `from __future__ import absolute_import` (no-absolute-import)
> 
> The `--py3k` flag for pylint enables _porting checks_. One of these is related to absolute imports (see [some more details](https://github.com/PyCQA/pylint/issues/2115#issuecomment-389810192)). For proper testing of python3 code (already ported) you should omit the `--py3k` flag.
> 
> > It still has python3 in shebang, breaking python2 environments. Probably this commit should be reverted for now: [a3319a6](https://github.com/munin-monitoring/contrib/commit/a3319a6dc92fc25dbe6f98eab684b702befbbfff)
> 
> I am sorry: due to python2 being end-of-life since the beginning of this year, I would strongly prefer to fix it for python3 instead.
> 

Yes, I think that's the plan to follow, making the plugin behave well under Python3

**Comment by ilf** [comment]
> I am sorry: due to python2 being end-of-life since the beginning of this year, I would strongly prefer to fix it for python3 instead.

This is not either-or.

I am very much for fixing the script to work well with Python 3, it's high time to do that.

But I would prefer not to explicitly exclude Python 2, as long as Distros like Debian still symlink /usr/bin/python to python2.
Also, enforcing a specific version might backfire in the future. What are we going to do when Python4 is released?

tl;dr: Let's use `/usr/bin/python` in shebang.

**Comment by sumpfralle** [comment]
> But I would prefer not to explicitly exclude Python 2, as long as Distros like Debian still symlink /usr/bin/python to python2.

As far as I understand the plan of the Python packaging team in Debian, there will be at least one Debian release *without* any binary providing `/usr/bin/python` (or maybe this path will be unpopulated forever). This is meant to force people to actively notice the problem instead of being surprised that the interpreter just behaves slightly different than before.

See https://lists.debian.org/debian-python/2017/12/msg00022.html.
> Also, enforcing a specific version might backfire in the future. What are we going to do when Python4 is released?

In case of a future incompatible major Python release, the plugin will need a deliberate switch to this version of the interpreter at some time.

Personally I am confident, that the end of support (as it happened now to python2) is the latest moment where a plugin can responsibly be considered *actively maintained*.

Users of python2 (without access to python3) are free to use the older version of the plugin (as it is still available in the git history).

Given that I just merged #1068 (thanks, @daftaupe), I think, that this issue can be closed, if no one objects.

**Comment by ilf** [comment]
> Given that I just merged #1068 (thanks, @daftaupe), I think, that this issue can be closed, if no one objects.

Thanks to https://github.com/munin-monitoring/contrib/pull/1068/commits/dabf2c644077151a0e36b28627bfcd9609505fde, the `dict.iter*()`-errors are fixed, but this one is still present:

```
% pylint --py3k tor_
************* Module tor_
tor_:58:0: W1618: import missing `from __future__ import absolute_import` (no-absolute-import)

------------------------------------------------------------------
Your code has been rated at 9.97/10 (previous run: 8.10/10, +1.87)
```

**Comment by sumpfralle** [comment]
> [..] but this one is still present:

I commented on this detail a few posts before:

>>    W1618: import missing from __future__ import absolute_import (no-absolute-import)
>
> The --py3k flag for pylint enables porting checks. One of these is related to absolute imports (see some more details). For proper testing of python3 code (already ported) you should omit the --py3k flag.

Thanks you all for the discussion and the resulting plugin improvement!

**Comment by sumpfralle** [comment]
From my point of view, the `tor_` plugin is in good shape now. Closing this one ...

**Comment by ilf** [comment]
> The `--py3k` flag for pylint enables _porting checks_. One of these is related to absolute imports (see [some more details](https://github.com/PyCQA/pylint/issues/2115#issuecomment-389810192)). For proper testing of python3 code (already ported) you should omit the `--py3k` flag.

`pylint --py3k tor_` only has one message:

```
tor_:74:0: W1618: import missing `from __future__ import absolute_import` (no-absolute-import)
```

`pylint tor_`gives a lot more:

```
************* Module tor_
tor_:97:0: W0622: Redefining built-in 'ConnectionError' (redefined-builtin)
tor_:89:0: C0103: Constant name "default_torcachefile" doesn't conform to UPPER_CASE naming style (invalid-name)
tor_:90:0: C0103: Constant name "default_torconnectmethod" doesn't conform to UPPER_CASE naming style (invalid-name)
tor_:91:0: C0103: Constant name "default_torgeoippath" doesn't conform to UPPER_CASE naming style (invalid-name)
tor_:92:0: C0103: Constant name "default_tormaxcountries" doesn't conform to UPPER_CASE naming style (invalid-name)
tor_:93:0: C0103: Constant name "default_torport" doesn't conform to UPPER_CASE naming style (invalid-name)
tor_:94:0: C0103: Constant name "default_torsocket" doesn't conform to UPPER_CASE naming style (invalid-name)
tor_:105:0: C0116: Missing function or method docstring (missing-function-docstring)
tor_:115:8: W0707: Consider explicitly re-raising using the 'from' keyword (raise-missing-from)
tor_:124:0: C0116: Missing function or method docstring (missing-function-docstring)
tor_:126:4: R1705: Unnecessary "elif" after "return" (no-else-return)
tor_:143:0: C0115: Missing class docstring (missing-class-docstring)
tor_:143:0: R0205: Class 'TorPlugin' inherits from object, can be safely removed from bases in python3 (useless-object-inheritance)
tor_:147:4: C0116: Missing function or method docstring (missing-function-docstring)
tor_:151:4: C0116: Missing function or method docstring (missing-function-docstring)
tor_:163:12: W0621: Redefining name 'stem' from outer scope (line 81) (redefined-outer-name)
tor_:169:12: W0621: Redefining name 'GeoIP' from outer scope (line 80) (redefined-outer-name)
tor_:161:4: C0116: Missing function or method docstring (missing-function-docstring)
tor_:163:12: C0415: Import outside toplevel (stem) (import-outside-toplevel)
tor_:165:8: C0103: Variable name "e" doesn't conform to snake_case naming style (invalid-name)
tor_:169:12: C0415: Import outside toplevel (GeoIP) (import-outside-toplevel)
tor_:171:8: C0103: Variable name "e" doesn't conform to snake_case naming style (invalid-name)
tor_:179:16: C0103: Variable name "e" doesn't conform to snake_case naming style (invalid-name)
tor_:169:12: W0611: Unused import GeoIP (unused-import)
tor_:186:4: C0116: Missing function or method docstring (missing-function-docstring)
tor_:193:4: C0116: Missing function or method docstring (missing-function-docstring)
tor_:202:0: C0115: Missing class docstring (missing-class-docstring)
tor_:203:4: W0231: __init__ method from base class 'TorPlugin' is not called (super-init-not-called)
tor_:220:12: C0103: Variable name "e" doesn't conform to snake_case naming style (invalid-name)
tor_:240:0: C0115: Missing class docstring (missing-class-docstring)
tor_:241:4: W0231: __init__ method from base class 'TorPlugin' is not called (super-init-not-called)
tor_:274:12: C0103: Variable name "e" doesn't conform to snake_case naming style (invalid-name)
tor_:278:0: C0115: Missing class docstring (missing-class-docstring)
tor_:279:4: W0231: __init__ method from base class 'TorPlugin' is not called (super-init-not-called)
tor_:304:12: C0103: Variable name "c" doesn't conform to snake_case naming style (invalid-name)
tor_:304:15: C0103: Variable name "v" doesn't conform to snake_case naming style (invalid-name)
tor_:311:51: C0103: Variable name "f" doesn't conform to snake_case naming style (invalid-name)
tor_:304:15: W0612: Unused variable 'v' (unused-variable)
tor_:319:46: C0103: Variable name "f" doesn't conform to snake_case naming style (invalid-name)
tor_:326:12: C0103: Variable name "c" doesn't conform to snake_case naming style (invalid-name)
tor_:326:15: C0103: Variable name "v" doesn't conform to snake_case naming style (invalid-name)
tor_:337:4: C0103: Argument name "cn" doesn't conform to snake_case naming style (invalid-name)
tor_:359:16: C0103: Variable name "c" doesn't conform to snake_case naming style (invalid-name)
tor_:361:12: C0103: Variable name "e" doesn't conform to snake_case naming style (invalid-name)
tor_:366:0: C0115: Missing class docstring (missing-class-docstring)
tor_:367:4: W0231: __init__ method from base class 'TorPlugin' is not called (super-init-not-called)
tor_:390:12: C0103: Variable name "e" doesn't conform to snake_case naming style (invalid-name)
tor_:394:0: C0115: Missing class docstring (missing-class-docstring)
tor_:395:4: W0231: __init__ method from base class 'TorPlugin' is not called (super-init-not-called)
tor_:412:12: C0103: Variable name "e" doesn't conform to snake_case naming style (invalid-name)
tor_:436:0: C0115: Missing class docstring (missing-class-docstring)
tor_:437:4: W0231: __init__ method from base class 'TorPlugin' is not called (super-init-not-called)
tor_:453:12: C0103: Variable name "e" doesn't conform to snake_case naming style (invalid-name)
tor_:470:0: C0115: Missing class docstring (missing-class-docstring)
tor_:471:4: W0231: __init__ method from base class 'TorPlugin' is not called (super-init-not-called)
tor_:489:12: C0103: Variable name "e" doesn't conform to snake_case naming style (invalid-name)
tor_:512:0: C0116: Missing function or method docstring (missing-function-docstring)
tor_:512:0: R0912: Too many branches (16/12) (too-many-branches)

------------------------------------------------------------------
Your code has been rated at 8.07/10 (previous run: 9.97/10, -1.90)
```

**Comment by sumpfralle** [comment]
> pylint --py3k tor_ only has one message:

The pylint author explains the [recommended usage of --py3k](https://github.com/PyCQA/pylint/issues/2115#issuecomment-389810192). Since this plugin is considered to be Python3 code now, it should not apply.

> pylint tor_ has a lot more:

We use `flake8` as the reference linter for the python-based plugins in this repository. `flake8` is way more tolerant than `pylint`.

Anyway: you are very welcome to propose a pull request, that improves the style of this plugin. `flake8` is just our minimal threshold for being acceptable.
